libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt

```diff
@@ -1505,6 +1527,67 @@ internal class MapRouteLine(
                 else -> y2
             }
         }
+
+        fun getRouteLineTrafficExpressionData(route: DirectionsRoute):
+            List<RouteLineTrafficExpressionData> {
+                val distances = route.legs()?.mapNotNull {
+                    it.annotation()?.distance()
+                }?.flatten()
+                ifNonNull(distances) { distanceList ->
+                    val roadClassMap = mutableMapOf<Int, String>()
+                    route.legs()?.asSequence()
+                        ?.map { it.steps() }
+                        ?.filterNotNull()
+                        ?.flatten()
+                        ?.mapNotNull { it.intersections() }
+                        ?.flatten()
+                        ?.filter {
+                            it.geometryIndex() != null && it.mapboxStreetsV8()?.roadClass() != null
+                        }?.toList()?.forEach {
+                            roadClassMap[it.geometryIndex()!!] =
+                                it.mapboxStreetsV8()!!.roadClass()!!
+                        }
+
+                    return route.legs()?.mapNotNull {
+                        it.annotation()?.congestion()
+                    }?.flatten()?.mapIndexed { index, congestion ->
+                        val distanceOffset = distanceList.subList(0, index).sum()
+                        RouteLineTrafficExpressionData(
+                            distanceOffset,
+                            congestion,
+                            roadClassMap[index]
+                        )
+                    } ?: listOf()
+                }
+                Timber.i(
+                    "Route leg annotation does not contain distance and/or congestion " +
+                        "so route line traffic cannot be calculated."
+                )
+                return listOf()
+            }
+
+        fun getRouteLineExpressionDataWithStreetClassOverride(
+            trafficExpressionData: List<RouteLineTrafficExpressionData>,
+            routeDistance: Double,
+            congestionColorProvider: (String, Boolean) -> Int,
+            isPrimaryRoute: Boolean,
+            trafficOverrideRoadClasses: List<String>
+        ): List<RouteLineExpressionData> {
+            return trafficExpressionData.map {
+                val percentDistanceTraveled = it.distanceFromOrigin / routeDistance
+                val trafficIdentifier =
+                    if (
+                        trafficOverrideRoadClasses.contains(it.roadClass ?: "unknown") &&
```

Comment:
There are cases I've come across where the road class for a given intersection is null.  I want to provide a way for a developer to indicate what should be done in this case.  My intention was that "unknown" could be included in the collection of the road classes defined in the styles.xml.  This would provide an option for making a substitution when the road class is null or not making a substitution. If I make the change you suggest there would be no option and instead the same behavior would be implemented for everyone.

Reply:
I think we should limit the scope of the feature to match the data that's returned by the Directions API. If the road class is unavailable, it means that it also isn't available in the map data so there will be no discrepancy in styling between the map and the route, which ultimately is what we're trying to resolve.

If we're trying to categorize unavailable road classes as `"unknown"`, I think that it should be done on the Directions API level.

Reply:
Fair enough. I'll make an adjustment.